### PR TITLE
Issue 2220

### PR DESCRIPTION
**Pull request by chebizarro**
This pull request addresses issue #2220 

It was relatively straight-forward to create a custom documenter for the AbstractCommandLine class but many of the applications needed modifications to the formatting of the docstrings of their dynamic attributes to be compatible with rst/docutils. The resulting changes render very well in html but result in some minor artifacts (escape characters) in the command line help text. 

- [X] I hereby agree to dual licence this and any previous contributions under both
the _Biopython License Agreement_ **AND** the _BSD 3-Clause License_.

- [X] I have read the ``CONTRIBUTING.rst`` file, have run ``flake8`` locally, and
understand that AppVeyor and TravisCI will be used to confirm the Biopython unit
tests and style checks pass with these changes.

- [X] I have added my name to the alphabetical contributors listings in the files
``NEWS.rst`` and ``CONTRIB.rst`` as part of this pull request, am listed
already, or do not wish to be listed. (*This acknowledgement is optional.*)

### CODE REVIEW

File: Bio/Blast/Applications.py [hunk 1]
```diff
@@ -968,14 +968,14 @@ def __init__(self, cmd="rpsblast", **kwargs):
         self.parameters = [
             # Query filtering options:
             _Option(["-seg", "seg"],
-                    """Filter query sequence with SEG (string).
+                    r"""Filter query sequence with SEG (string).
 
                     Format: "yes", "window locut hicut", or "no" to disable.
-                    Default is "12 2.2 2.5""",
+                    Default is "12 2.2 2.5" """,
```
Comment (peterjc) [inline Bio/Blast/Applications.py]: Style wise that closing triple quote ought to get its own line.

File: Bio/Blast/Applications.py [hunk 2]
```diff
@@ -715,14 +715,14 @@ def __init__(self, cmd="tblastn", **kwargs):
...
+                    r"""Use composition-based statistics (string, default 2, i.e. True).
 
                     0, F or f: no composition-based statistics
                     1: Composition-based statistics as in NAR 29:2994-3005, 2001
                     2, T or t, D or d : Composition-based score adjustment as in
-                       Bioinformatics 21:902-911, 2005, conditioned on sequence properties
+                    \   Bioinformatics 21:902-911, 2005, conditioned on sequence properties
                     3: Composition-based score adjustment as in Bioinformatics 21:902-911,
-                       2005, unconditionally
+                    \   2005, unconditionally
```
Comment (peterjc) [inline Bio/Blast/Applications.py]: Why this slash here and on the second bullet point? These docstrings should be valid RST.

File: Bio/Blast/Applications.py [hunk 3]
```diff
@@ -984,26 +984,26 @@ def __init__(self, cmd="rpsblast", **kwargs):
...
                     equate=False),
             # General search options:
             _Option(["-comp_based_stats", "comp_based_stats"],
-                    """Use composition-based statistics.
+                    r"""Use composition-based statistics.
 
                     D or d: default (equivalent to 0 )
                     0 or F or f: Simplified Composition-based statistics as in
-                                 Bioinformatics 15:1000-1011, 1999
+                    \             Bioinformatics 15:1000-1011, 1999
```
Comment (peterjc) [inline Bio/Blast/Applications.py]: Again why the slash. I think the simplest way to make this look nice as RST is to make each of these into its own paragraph by adding blank lines?

**Comment by peterjc** [comment]
I didn't even know this was possible with Sphinx! Nice 👍 

I have a repeated query about slashes on continued lines in the docstrings - many of which can likely be edited in multiple ways to produce nicer RST output.

File: Bio/Blast/Applications.py (hunk 3 above)
Comment (chebizarro) [reply Bio/Blast/Applications.py]: I suspect my IDE was inserting extraneous white space and triggering W293. I've amended those particular files and reformatted them for nicer output.

File: Bio/Sequencing/Applications/_bwa.py [hunk 4]
```diff
@@ -431,7 +431,7 @@ def __init__(self, cmd="bwa", **kwargs):
                     checker_function=lambda x: isinstance(x, int),
                     equate=False),
             _Option(["-d", "d"],
-                    "Off-diagonal X-dropoff (Z-dropoff). Stop extension when the difference between the best and the current extension score is above |i-j|*A+INT, where i and j are the current positions of the query and reference, respectively, and A is the matching score. Z-dropoff is similar to BLAST\'s X-dropoff except that it doesn\'t penalize gaps in one of the sequences in the alignment. Z-dropoff not only avoids unnecessary extension, but also reduces poor alignments inside a long good alignment. [100]",
+                    r"Off-diagonal X-dropoff (Z-dropoff). Stop extension when the difference between the best and the current extension score is above \|i-j\|*A+INT, where i and j are the current positions of the query and reference, respectively, and A is the matching score. Z-dropoff is similar to BLAST\'s X-dropoff except that it doesn\'t penalize gaps in one of the sequences in the alignment. Z-dropoff not only avoids unnecessary extension, but also reduces poor alignments inside a long good alignment. [100]",
```
Comment (peterjc) [inline Bio/Sequencing/Applications/_bwa.py]: We don't need to slash escape the apostrophes here do we?

**Comment by peterjc** [comment]
Still some slashes in `` Bio/Phylo/Applications/_Phyml.py`` and `` Bio/Phylo/Applications/_Raxml.py`` - good to know it was just an editor doing something silly.

**Comment by peterjc** [comment]
This is a big improvement, although running it locally I see something odd. e.g.

```
$ tox -c .travis-tox.ini -e api
$ open Doc/api/_build/html/Bio.Blast.Applications.html
```

Everything seems to be listed twice per class, once as attributes of the object, and again (after the methods) as properties. Visually the property list looks nicer, so I'd hide the attributes given the choice. This could just be how properties are displayed by default?

Do you see the same thing (might depend on versions of Sphinx etc)?

If you don't immediately see how to fix this, I am inclined to merge this as it is still an improvement.

**Comment by chebizarro** [comment]
The latest commit should fix this - the duplicate attributes are apparently caused by numpydoc clashing with autosummary.

**Comment by peterjc** [comment]
Thank you - confirmed after testing locally. This should be live here soon:
https://biopython.org/docs/dev/api/

**Pull request merged**